File: mybox/package/installer/apt.py
from typing import Optional

from .base import Installer


class Apt(Installer):
    async def install(self, package: str) -> None:
        await self.driver.with_root(True).run(
            "env", "DEBIAN_FRONTEND=noninteractive", "apt", "install", "--yes", package
        )

    async def upgrade(self, package: str) -> None:
        await self.install(package)

    async def latest_version(self, package: str) -> str:
        result = await self.driver.run_output_(
            "apt-cache",
            "show",
            "--quiet",
            "--no-all-versions",
            package,
        )
        if not result.ok:
            raise ValueError(f"Cannot determine version for: {package}.")
        for line in result.output.splitlines():
            line = line.strip()
            if line.startswith("Version:"):
                return line.split(": ", 1)[-1]
        raise Exception(f"Cannot parse apt output: {result.output}")  # pragma: no cover

    async def installed_version(self, package: str) -> Optional[str]:
        return (
            await self.driver.run_output_(
                "dpkg-query", "--showformat", "${Version}", "--show", package
            )
        ).output

File: src/Mybox/Installer/Apt.hs
module Mybox.Installer.Apt (apt) where

import Data.Map.Strict qualified as Map
import Data.Text qualified as Text

import Mybox.Driver
import Mybox.Installer.Class
import Mybox.Prelude
import Mybox.Stores

aptInstall :: Driver :> es => Text -> Eff es ()
aptInstall package =
  drvRun $
    "sudo"
      :| [ "env"
         , "DEBIAN_FRONTEND=noninteractive"
         , "apt"
         , "install"
         , "-y"
         , package
         ]

parseAptCacheOutput :: Text -> Maybe Text
parseAptCacheOutput output = listToMaybe $ do
  line <- Text.lines output
  toList $ Text.stripPrefix "Version: " line

aptInstalled :: Driver :> es => Text -> Eff es (Maybe Text)
aptInstalled package = do
  result <-
    drvRunOutputExit $
      "dpkg-query"
        :| [ "--showformat"
           , "${Version}"
           , "--show"
           , package
           ]
  pure $ if result.exit == ExitSuccess && not (Text.null result.output) then Just result.output else Nothing

aptLatest :: Driver :> es => Text -> Eff es Text
aptLatest package = do
  output <-
    drvRunOutput $
      "apt-cache"
        :| [ "show"
           , "--quiet"
           , "--no-all-versions"
           , package
           ]
  maybe (terror $ "Cannot determine version for: " <> package) pure $ parseAptCacheOutput output

aptPackageInfo :: Driver :> es => Maybe Text -> Eff es (Map Text PackageVersion)
aptPackageInfo Nothing = pure mempty
aptPackageInfo (Just package) = fmap (Map.singleton package) $ PackageVersion <$> aptInstalled package <*> aptLatest package

apt :: Installer
apt =
  Installer
    { storePackages = jsonStore "apt"
    , storeGlobal = jsonStore "apt-global"
    , install_ = aptInstall
    , upgrade_ = aptInstall
    , getPackageInfo = aptPackageInfo
    }
